File: lib/idp_common_pkg/idp_common/config/schema_constants.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0

"""
Constants for AWS IDP-specific JSON Schema extensions.

These custom fields extend standard JSON Schema to support IDP-specific features
like document types, attribute types, and evaluation methods.
"""

# ============================================================================
# JSON Schema Standard Fields
# ============================================================================
SCHEMA_FIELD = "$schema"
ID_FIELD = "$id"
DEFS_FIELD = "$defs"
REF_FIELD = "$ref"

# ============================================================================
# AWS IDP Document Type Extensions
# ============================================================================
# Marks a schema as a document type (top-level class)
X_AWS_IDP_DOCUMENT_TYPE = "x-aws-idp-document-type"

# Classification metadata for document type
X_AWS_IDP_CLASSIFICATION = "x-aws-idp-classification"

# Regex patterns for classification optimization
X_AWS_IDP_DOCUMENT_NAME_REGEX = "x-aws-idp-document-name-regex"
X_AWS_IDP_PAGE_CONTENT_REGEX = "x-aws-idp-document-page-content-regex"

# ============================================================================
# Legacy Attribute Type Values (for migration only)
# ============================================================================
# These map to standard JSON Schema types:
# - "simple" → type: "string"
# - "group" → type: "object" with properties
# - "list" → type: "array" with items
ATTRIBUTE_TYPE_SIMPLE = "simple"
ATTRIBUTE_TYPE_GROUP = "group"
ATTRIBUTE_TYPE_LIST = "list"

# ============================================================================
# AWS IDP List-Specific Extensions
# ============================================================================
# Description for list items
X_AWS_IDP_LIST_ITEM_DESCRIPTION = "x-aws-idp-list-item-description"

# Original attribute name (preserved from legacy format)
X_AWS_IDP_ORIGINAL_NAME = "x-aws-idp-original-name"

# ============================================================================
# AWS IDP Evaluation Extensions
# ============================================================================
# Evaluation method for attribute comparison
X_AWS_IDP_EVALUATION_METHOD = "x-aws-idp-evaluation-method"


X_AWS_IDP_EXAMPLES = "x-aws-idp-examples"

# Valid evaluation methods
EVALUATION_METHOD_EXACT = "EXACT"
EVALUATION_METHOD_NUMERIC_EXACT = "NUMERIC_EXACT"
EVALUATION_METHOD_FUZZY = "FUZZY"
EVALUATION_METHOD_SEMANTIC = "SEMANTIC"
EVALUATION_METHOD_LLM = "LLM"

VALID_EVALUATION_METHODS = frozenset(
    [
        EVALUATION_METHOD_EXACT,
        EVALUATION_METHOD_NUMERIC_EXACT,
        EVALUATION_METHOD_FUZZY,
        EVALUATION_METHOD_SEMANTIC,
        EVALUATION_METHOD_LLM,
    ]
)

# Confidence threshold for evaluation (0.0 to 1.0)
X_AWS_IDP_CONFIDENCE_THRESHOLD = "x-aws-idp-confidence-threshold"

# ============================================================================
# AWS IDP Prompt Extensions
# ============================================================================
# Custom prompt override for attribute extraction
X_AWS_IDP_PROMPT_OVERRIDE = "x-aws-idp-prompt-override"

# Maximum length for prompt overrides
MAX_PROMPT_OVERRIDE_LENGTH = 10000

# ============================================================================
# AWS IDP Example Extensions
# ============================================================================
# Extensions for few-shot example support
X_AWS_IDP_CLASS_PROMPT = "x-aws-idp-class-prompt"
X_AWS_IDP_ATTRIBUTES_PROMPT = "x-aws-idp-attributes-prompt"
X_AWS_IDP_IMAGE_PATH = "x-aws-idp-image-path"

# ============================================================================
# Legacy Format Field Names
# ============================================================================
# These are the field names used in the legacy format
LEGACY_ATTRIBUTES = "attributes"
LEGACY_NAME = "name"
LEGACY_DESCRIPTION = "description"
LEGACY_ATTRIBUTE_TYPE = "attributeType"
LEGACY_GROUP_ATTRIBUTES = "groupAttributes"
LEGACY_LIST_ITEM_TEMPLATE = "listItemTemplate"
LEGACY_ITEM_ATTRIBUTES = "itemAttributes"
LEGACY_ITEM_DESCRIPTION = "itemDescription"
LEGACY_EVALUATION_METHOD = "evaluation_method"
LEGACY_CONFIDENCE_THRESHOLD = "confidence_threshold"
LEGACY_PROMPT_OVERRIDE = "prompt_override"

# Legacy example fields
LEGACY_EXAMPLES = "examples"
LEGACY_CLASS_PROMPT = "classPrompt"
LEGACY_ATTRIBUTES_PROMPT = "attributesPrompt"
LEGACY_IMAGE_PATH = "imagePath"

# Legacy regex fields (same name in both legacy and new format)
LEGACY_DOCUMENT_NAME_REGEX = "document_name_regex"
LEGACY_DOCUMENT_PAGE_CONTENT_REGEX = "document_page_content_regex"

# ============================================================================
# JSON Schema Standard Property Names
# ============================================================================
SCHEMA_TYPE = "type"
SCHEMA_PROPERTIES = "properties"
SCHEMA_ITEMS = "items"
SCHEMA_REQUIRED = "required"
SCHEMA_DESCRIPTION = "description"

# JSON Schema type values
TYPE_OBJECT = "object"
TYPE_ARRAY = "array"
TYPE_STRING = "string"
TYPE_NUMBER = "number"
TYPE_INTEGER = "integer"
TYPE_BOOLEAN = "boolean"
TYPE_NULL = "null"

File: src/ui/src/constants/schemaConstants.js
/**
 * JSON Schema and AWS IDP Extension Constants
 *
 * These constants match the backend lib/idp_common_pkg/idp_common/config/schema_constants.py
 * to ensure consistency between frontend and backend when building and processing document schemas.
 *
 * @see lib/idp_common_pkg/idp_common/config/schema_constants.py
 */

// ============================================================================
// JSON Schema Standard Fields
// ============================================================================
export const SCHEMA_FIELD = '$schema';
export const ID_FIELD = '$id';
export const DEFS_FIELD = '$defs';
export const REF_FIELD = '$ref';

// ============================================================================
// JSON Schema Standard Property Names
// ============================================================================
export const SCHEMA_TYPE = 'type';
export const SCHEMA_PROPERTIES = 'properties';
export const SCHEMA_ITEMS = 'items';
export const SCHEMA_REQUIRED = 'required';
export const SCHEMA_DESCRIPTION = 'description';
export const SCHEMA_EXAMPLES = 'examples';

// ============================================================================
// JSON Schema Type Values
// ============================================================================
export const TYPE_OBJECT = 'object';
export const TYPE_ARRAY = 'array';
export const TYPE_STRING = 'string';
export const TYPE_NUMBER = 'number';
export const TYPE_INTEGER = 'integer';
export const TYPE_BOOLEAN = 'boolean';
export const TYPE_NULL = 'null';

// UI-friendly type options for dropdowns
export const TYPE_OPTIONS = [
  { label: 'String', value: TYPE_STRING },
  { label: 'Number', value: TYPE_NUMBER },
  { label: 'Integer', value: TYPE_INTEGER },
  { label: 'Boolean', value: TYPE_BOOLEAN },
  { label: 'Object', value: TYPE_OBJECT },
  { label: 'Array', value: TYPE_ARRAY },
  { label: 'Null', value: TYPE_NULL },
];

// Type colors for UI visualization
export const TYPE_COLORS = {
  [TYPE_STRING]: 'blue',
  [TYPE_NUMBER]: 'green',
  [TYPE_INTEGER]: 'green',
  [TYPE_BOOLEAN]: 'grey',
  [TYPE_OBJECT]: 'red',
  [TYPE_ARRAY]: 'purple',
  [TYPE_NULL]: 'grey',
};

// ============================================================================
// AWS IDP Document Type Extensions
// ============================================================================
/** Marks a schema as a document type (top-level class) */
export const X_AWS_IDP_DOCUMENT_TYPE = 'x-aws-idp-document-type';

/** Classification metadata for document type */
export const X_AWS_IDP_CLASSIFICATION = 'x-aws-idp-classification';

/** Regex patterns for classification optimization */
export const X_AWS_IDP_DOCUMENT_NAME_REGEX = 'x-aws-idp-document-name-regex';
export const X_AWS_IDP_PAGE_CONTENT_REGEX = 'x-aws-idp-document-page-content-regex';

// ============================================================================
// AWS IDP List-Specific Extensions
// ============================================================================
/** Description for list items */
export const X_AWS_IDP_LIST_ITEM_DESCRIPTION = 'x-aws-idp-list-item-description';

/** Original attribute name (preserved from legacy format) */
export const X_AWS_IDP_ORIGINAL_NAME = 'x-aws-idp-original-name';

// ============================================================================
// AWS IDP Evaluation Extensions (Stickler-based Baseline Accuracy)
// ============================================================================
/** Evaluation method for baseline comparison */
export const X_AWS_IDP_EVALUATION_METHOD = 'x-aws-idp-evaluation-method';

/** Evaluation threshold for baseline match (0.0 to 1.0) */
export const X_AWS_IDP_EVALUATION_THRESHOLD = 'x-aws-idp-evaluation-threshold';

/** Evaluation weight for field importance (business criticality) */
export const X_AWS_IDP_EVALUATION_WEIGHT = 'x-aws-idp-evaluation-weight';

/** Overall match threshold for document-level evaluation (0.0 to 1.0) */
export const X_AWS_IDP_EVALUATION_MATCH_THRESHOLD = 'x-aws-idp-evaluation-match-threshold';

/** Evaluation model name (optional, defaults to document type name) */
export const X_AWS_IDP_EVALUATION_MODEL_NAME = 'x-aws-idp-evaluation-model-name';

// Valid evaluation methods
export const EVALUATION_METHOD_EXACT = 'EXACT';
export const EVALUATION_METHOD_NUMERIC_EXACT = 'NUMERIC_EXACT';
export const EVALUATION_METHOD_FUZZY = 'FUZZY';
export const EVALUATION_METHOD_LEVENSHTEIN = 'LEVENSHTEIN';
export const EVALUATION_METHOD_SEMANTIC = 'SEMANTIC';
export const EVALUATION_METHOD_LLM = 'LLM';
export const EVALUATION_METHOD_HUNGARIAN = 'HUNGARIAN';

export const VALID_EVALUATION_METHODS = [
  EVALUATION_METHOD_EXACT,
  EVALUATION_METHOD_NUMERIC_EXACT,
  EVALUATION_METHOD_FUZZY,
  EVALUATION_METHOD_LEVENSHTEIN,
  EVALUATION_METHOD_SEMANTIC,
  EVALUATION_METHOD_LLM,
  EVALUATION_METHOD_HUNGARIAN,
];

// UI-friendly evaluation method options with descriptions and validation metadata
export const EVALUATION_METHOD_OPTIONS = [
  {
    label: 'Exact',
    value: EVALUATION_METHOD_EXACT,
    description: 'Character-by-character match after normalization',
    validFor: [TYPE_STRING, TYPE_NUMBER, TYPE_INTEGER, TYPE_BOOLEAN],
  },
  {
    label: 'Numeric Exact',
    value: EVALUATION_METHOD_NUMERIC_EXACT,
    description: 'Numeric comparison (currency/format normalized)',
    validFor: [TYPE_NUMBER, TYPE_INTEGER, TYPE_STRING],
  },
  {
    label: 'Fuzzy',
    value: EVALUATION_METHOD_FUZZY,
    description: 'Allows minor formatting variations',
    validFor: [TYPE_STRING],
  },
  {
    label: 'Levenshtein',
    value: EVALUATION_METHOD_LEVENSHTEIN,
    description: 'String distance-based matching',
    validFor: [TYPE_STRING],
  },
  {
    label: 'Semantic',
    value: EVALUATION_METHOD_SEMANTIC,
    description: 'Embedding-based meaning comparison',
    validFor: [TYPE_STRING, TYPE_OBJECT],
  },
  {
    label: 'LLM',
    value: EVALUATION_METHOD_LLM,
    description: 'AI-powered functional equivalence',
    validFor: [TYPE_STRING, TYPE_OBJECT],
  },
  {
    label: 'Hungarian',
    value: EVALUATION_METHOD_HUNGARIAN,
    description: 'Optimal matching for arrays of objects (List[Object] only)',
    validFor: [TYPE_ARRAY],
    requiresStructuredItems: true, // Only for arrays with object items
  },
];

// Default threshold values per evaluation method (for regular fields)
// Only methods that use similarity thresholds have defaults here
export const EVALUATION_THRESHOLD_DEFAULTS = {
  [EVALUATION_METHOD_FUZZY]: 0.85,
  [EVALUATION_METHOD_LEVENSHTEIN]: 0.8,
  [EVALUATION_METHOD_SEMANTIC]: 0.7,
  // No defaults for EXACT (binary), NUMERIC_EXACT (uses tolerance), LLM (binary)
};

// Default match_threshold values per evaluation method (for structured arrays)
// Note: Only HUNGARIAN uses match_threshold. LLM evaluates arrays semantically without match_threshold.
export const EVALUATION_MATCH_THRESHOLD_DEFAULTS = {
  [EVALUATION_METHOD_HUNGARIAN]: 0.8,
};

// Methods that require threshold configuration (for non-array fields)
// Only methods that use similarity-based scoring require thresholds
export const METHODS_REQUIRING_THRESHOLD = [
  EVALUATION_METHOD_FUZZY,
  EVALUATION_METHOD_LEVENSHTEIN,
  EVALUATION_METHOD_SEMANTIC,
  // EXACT, NUMERIC_EXACT, and LLM do not use thresholds
];

// Methods that require match_threshold configuration (for structured arrays)
// Note: Only HUNGARIAN uses match_threshold for item-by-item optimal matching.
// LLM evaluates arrays semantically as a whole and does not use match_threshold.
export const METHODS_REQUIRING_MATCH_THRESHOLD = [EVALUATION_METHOD_HUNGARIAN];

// ============================================================================
// AWS IDP Assessment Extensions (Extraction Confidence Alerts)
// ============================================================================
/** Confidence threshold for extraction quality alerts (0.0 to 1.0) */
export const X_AWS_IDP_CONFIDENCE_THRESHOLD = 'x-aws-idp-confidence-threshold';

// ============================================================================
// AWS IDP Prompt Extensions
// ============================================================================
/** Custom prompt override for attribute extraction */
export const X_AWS_IDP_PROMPT_OVERRIDE = 'x-aws-idp-prompt-override';

/** Maximum length for prompt overrides */
export const MAX_PROMPT_OVERRIDE_LENGTH = 10000;

// ============================================================================
// AWS IDP Example Extensions
// ============================================================================
/** Extensions for few-shot example support */
export const X_AWS_IDP_EXAMPLES = 'x-aws-idp-examples';
export const X_AWS_IDP_CLASS_PROMPT = 'x-aws-idp-class-prompt';
export const X_AWS_IDP_ATTRIBUTES_PROMPT = 'x-aws-idp-attributes-prompt';
export const X_AWS_IDP_IMAGE_PATH = 'x-aws-idp-image-path';

// ============================================================================
// Legacy Format Field Names (for migration reference)
// ============================================================================
export const LEGACY_ATTRIBUTES = 'attributes';
export const LEGACY_NAME = 'name';
export const LEGACY_DESCRIPTION = 'description';
export const LEGACY_ATTRIBUTE_TYPE = 'attributeType';
export const LEGACY_GROUP_ATTRIBUTES = 'groupAttributes';
export const LEGACY_LIST_ITEM_TEMPLATE = 'listItemTemplate';
export const LEGACY_ITEM_ATTRIBUTES = 'itemAttributes';
export const LEGACY_ITEM_DESCRIPTION = 'itemDescription';

// Legacy attribute type values (deprecated - use JSON Schema types instead)
export const ATTRIBUTE_TYPE_SIMPLE = 'simple';
export const ATTRIBUTE_TYPE_GROUP = 'group';
export const ATTRIBUTE_TYPE_LIST = 'list';

// UI-friendly legacy attribute options (for migration UI)
export const ATTRIBUTE_TYPE_OPTIONS = [
  { label: 'Simple', value: ATTRIBUTE_TYPE_SIMPLE, description: 'Single value field' },
  { label: 'Group', value: ATTRIBUTE_TYPE_GROUP, description: 'Nested object with properties' },
  { label: 'List', value: ATTRIBUTE_TYPE_LIST, description: 'Array of items' },
];

// ============================================================================
// JSON Schema Format Options (with examples for UI display)
// See: https://json-schema.org/understanding-json-schema/reference/string#built-in-formats
// ============================================================================
export const FORMAT_OPTIONS = [
  { label: 'None', value: '', description: 'No format validation' },
  { label: 'Date (e.g., 2024-12-25)', value: 'date', description: 'ISO 8601 date: YYYY-MM-DD' },
  { label: 'Time (e.g., 14:30:00)', value: 'time', description: 'ISO 8601 time: HH:MM:SS' },
  { label: 'Date-Time (e.g., 2024-12-25T14:30:00Z)', value: 'date-time', description: 'ISO 8601 date-time' },
  { label: 'Duration (e.g., P3Y6M4D)', value: 'duration', description: 'ISO 8601 duration' },
  { label: 'Email (e.g., user@example.com)', value: 'email', description: 'RFC 5321 email address' },
  { label: 'IDN Email (e.g., user@example.com)', value: 'idn-email', description: 'Internationalized email' },
  { label: 'Hostname (e.g., example.com)', value: 'hostname', description: 'RFC 1123 hostname' },
  { label: 'IDN Hostname (e.g., example.com)', value: 'idn-hostname', description: 'Internationalized hostname' },
  { label: 'IPv4 (e.g., 192.168.1.1)', value: 'ipv4', description: 'IPv4 address' },
  { label: 'IPv6 (e.g., 2001:db8::1)', value: 'ipv6', description: 'IPv6 address' },
  { label: 'URI (e.g., https://example.com/path)', value: 'uri', description: 'RFC 3986 URI' },
  { label: 'URI Reference (e.g., /path/to/resource)', value: 'uri-reference', description: 'URI or relative reference' },
  { label: 'IRI (e.g., https://example.com/path)', value: 'iri', description: 'Internationalized URI' },
  { label: 'IRI Reference', value: 'iri-reference', description: 'IRI or relative reference' },
  { label: 'URI Template (e.g., /users/{id})', value: 'uri-template', description: 'RFC 6570 URI template' },
  { label: 'JSON Pointer (e.g., /foo/bar)', value: 'json-pointer', description: 'RFC 6901 JSON pointer' },
  { label: 'Relative JSON Pointer (e.g., 1/foo)', value: 'relative-json-pointer', description: 'Relative JSON pointer' },
  { label: 'Regex (e.g., ^[a-z]+$)', value: 'regex', description: 'ECMA-262 regular expression' },
  { label: 'UUID (e.g., 550e8400-e29b-41d4-a716-446655440000)', value: 'uuid', description: 'RFC 4122 UUID' },
];

// ============================================================================
// Content Encoding Options
// ============================================================================
export const CONTENT_ENCODING_OPTIONS = [
  { label: 'None', value: '' },
  { label: 'Base64', value: 'base64' },
  { label: '7bit', value: '7bit' },
  { label: '8bit', value: '8bit' },
  { label: 'Binary', value: 'binary' },
  { label: 'Quoted-Printable', value: 'quoted-printable' },
];

// ============================================================================
// Default Values
// ============================================================================
/** Default JSON Schema version */
export const DEFAULT_SCHEMA_VERSION = 'https://json-schema.org/draft/2020-12/schema';

/** Default confidence threshold */
export const DEFAULT_CONFIDENCE_THRESHOLD = 0.9;
